Add bitwise distributed reduction ops

pietern opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#26824 Add bitwise distributed reduction ops**

These ops are named after the bitwise reduction ops in MPI.

This is based on the work done by @knottb in #22449.

Closes #22449.

Differential Revision: [D17600210](https://our.internmc.facebook.com/intern/diff/D17600210)

pietern committed:
Add bitwise distributed reduction ops

These ops are named after the bitwise reduction ops in MPI.

This is based on the work done by @knottb in #22449.

Closes #22449.

pietern committed:
Update on "Add bitwise distributed reduction ops"


These ops are named after the bitwise reduction ops in MPI.

This is based on the work done by @knottb in #22449.

Closes #22449.

rohan-varma commented on torch/lib/c10d/ProcessGroupGloo.cpp:
```diff
@@ -113,7 +115,80 @@ class GlooStore : public ::gloo::rendezvous::Store {
+  auto tc = static_cast<T*>(c);
+  auto ta = static_cast<const T*>(a);
+  auto tb = static_cast<const T*>(b);
+  for (size_t i = 0; i < n; i++) {
```
Do we need to validate that we won't get indexing errors here, or does Gloo's reduce function provide us with the correct `n`?

rohan-varma reviewed (approved):
Do we need to validate that we won't get indexing errors in the implementation of band, bor, bxor, or does Gloo's reduce function provide us with the correct `n`?

zhaojuanmao reviewed (approved):
lgtm, too

pietern committed:
Update on "Add bitwise distributed reduction ops"


These ops are named after the bitwise reduction ops in MPI.

This is based on the work done by @knottb in #22449.

Closes #22449.

pietern commented:
@rohan-varma This is guaranteed to be in range, as a subset of the total number of elements.
